### PR TITLE
feat(pageserver): use split layer writer in gc-compaction

### PR DESCRIPTION
**Pull request by skyzh**
## Problem

Part of #8002, the final big PR in the batch.

## Summary of changes

This pull request uses the new split layer writer in the gc-compaction.

* It changes how layers are split. Previously, we split layers based on the original split point, but this creates too many layers (test_gc_feedback has one key per layer).
  * Therefore, we first verify if the layer map can be processed by the current algorithm (See https://github.com/neondatabase/neon/pull/8191, it's basically the same check)
  * On that, we proceed with the compaction. This way, it creates a large enough layer close to the target layer size.
* Added a new set of functions `with_discard` in the split layer writer. This helps us skip layers if we are going to produce the same persistent key.
* The delta writer will keep the updates of the same key in a single file. This might create a super large layer, but we can optimize it later.
* The split layer writer is used in the gc-compaction algorithm, and it will split layers based on size.
* Fix the image layer summary block encoded the wrong key range.

## Checklist before requesting a review

- [x] I have performed a self-review of my code.
- [x] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics? if so did you add the relevant metrics to the dashboard?
- [ ] If this PR requires public announcement, mark it with /release-notes label and add several sentences in this section.

## Checklist before merging

- [ ] Do not forget to reformat commit message to not include the above checklist

### CODE REVIEW

File: pageserver/src/tenant/timeline/compaction.rs [hunk 1]
```diff
@@ -2022,47 +1989,12 @@ impl Timeline {
...
 
         // Actually, we can decide not to write to the image layer at all at this point because
         // the key and LSN range are determined. However, to keep things simple here, we still
         // create this writer, and discard the writer in the end.
 
-        let mut delta_values = Vec::new();
-        let delta_split_points = delta_split_points.into_iter().collect_vec();
-        let mut current_delta_split_point = 0;
-        let mut delta_layers = Vec::new();
+        // let mut delta_layers = Vec::new();
```
Comment (skyzh) [inline pageserver/src/tenant/timeline/compaction.rs:1997]: will remove this line after CI completes, I'm waiting for the result from the benchmark

File: pageserver/src/tenant/storage_layer/split_writer.rs [hunk 2]
```diff
@@ -76,49 +107,108 @@ impl SplitImageLayerWriter {
...
         Ok((self.generated_layers, self.inner))
     }
 }
 
 /// A delta writer that takes key-lsn-values and produces multiple delta layers. The interface does not
 /// guarantee atomicity (i.e., if the delta layer generation fails, there might be leftover files
 /// to be cleaned up).
+///
+/// Note that if updates of a single key exceed the target size limit, all of the updates will be batched
+/// in a single file. This behavior might change in the future. For reference, the legacy compaction algorithm
```
Comment (arpad-m) [inline pageserver/src/tenant/storage_layer/split_writer.rs:199]:
```suggestion
/// into a single file. This behavior might change in the future. For reference, the legacy compaction algorithm
```

File: pageserver/src/tenant/storage_layer/split_writer.rs [hunk 3]
```diff
@@ -147,67 +237,141 @@ impl SplitDeltaLayerWriter {
...
+                        .push(SplitWriterResult::Discarded(layer_key));
+                } else {
+                    let (desc, path) = prev_delta_writer.finish(key, ctx).await?;
+                    let delta_layer = Layer::finish_creating(self.conf, tline, desc, &path)?;
+                    self.generated_layers
+                        .push(SplitWriterResult::Produced(delta_layer));
+                }
+            } else if self.inner.estimated_size() >= u32::MAX as u64 {
+                // We have to produce a very large file b/c a key is updated too often.
+                anyhow::bail!(
```
Comment (arpad-m) [inline pageserver/src/tenant/storage_layer/split_writer.rs:298]: We have a check against `S3_UPLOAD_LIMIT` a bit later, which I think is slightly above `u32::MAX`. We can of course limit it by `u32::MAX` as well, but not sure. Ideally we'd have an unified constant at least, or maybe this check is not needed, idk.

File: pageserver/src/tenant/timeline.rs [hunk 4]
```diff
@@ -5451,17 +5451,21 @@ impl Timeline {
...
-            let guard = self.layers.read().await;
-            for layer in guard.layer_map()?.iter_historic_layers() {
-                if layer.is_delta()
-                    && overlaps_with(&layer.lsn_range, &deltas.lsn_range)
-                    && layer.lsn_range != deltas.lsn_range
-                {
-                    // If a delta layer overlaps with another delta layer AND their LSN range is not the same, panic
-                    panic!(
+            if deltas.key_range.start.next() != deltas.key_range.end {
+                // skip single-key layer files
```
Comment (arpad-m) [inline pageserver/src/tenant/timeline.rs:5455]: why should we skip them? we still want the lsn ranges to not overlap.

File: pageserver/src/tenant/timeline/compaction.rs [hunk 5]
```diff
@@ -71,15 +74,61 @@ pub(crate) struct KeyHistoryRetention {
...
+    ///
+    /// Note that bottom-most-gc-compaction never _adds_ new data in that case, only removes.
+    ///
+    /// `discard_key` will only be called when the writer reaches its target (instead of for every key), so it's fine to grab a lock inside.
+    async fn discard_key(key: &PersistentLayerKey, tline: &Arc<Timeline>, dry_run: bool) -> bool {
+        if dry_run {
+            return true;
+        }
+        let guard = tline.layers.read().await;
+        if guard.contains_key(key) {
```
Comment (arpad-m) [inline pageserver/src/tenant/timeline/compaction.rs:94]:
```Rust
if !guard.contains_key(key) {
    return false;
}
```

File: pageserver/src/tenant/storage_layer/split_writer.rs [hunk 6]
```diff
@@ -147,67 +237,141 @@ impl SplitDeltaLayerWriter {
             timeline_id,
             tenant_shard_id,
             lsn_range,
+            last_key_written: Key::MIN,
+            start_key,
         })
     }
 
-    pub async fn put_value(
+    /// Put value into the layer writer. When the writer decides to produce a layer, and the discard fn returns true,
```
Comment (arpad-m) [inline pageserver/src/tenant/storage_layer/split_writer.rs:245]:
```suggestion
    /// Put value into the layer writer. In the case the writer decides to produce a layer, and the discard fn returns true, no layer will be written in the end.
```

**Review by arpad-m — CHANGES_REQUESTED** [review]
Mostly okay, although I'm not sure about the issue with the many updates to a single key. If we have such cases, they'd get major regressions (and we had them in the past due to aux files, so now we don't have them any more but who knows maybe there is other sources).

File: pageserver/src/tenant.rs [hunk 7]
```diff
@@ -7013,10 +7013,14 @@ mod tests {
...
                     is_delta: false
                 },
-                // The delta layer that is cut in the middle
+                // The delta layer covers the full range
                 PersistentLayerKey {
-                    key_range: get_key(3)..get_key(4),
-                    lsn_range: Lsn(0x30)..Lsn(0x41),
+                    key_range: {
+                        let mut key = Key::MAX;
+                        key.field6 -= 1;
```
[the remark this answers is not in the document]
Comment (koivunej) [reply pageserver/src/tenant.rs:7020]: Without any context from earlier PRs this looks like this range would contradict with the comment above.

[the remark this answers is not in the document]
Comment (arpad-m) [reply pageserver/src/tenant.rs:7020]: It does cover the full range, we just make it artificially smaller to avoid it being mistaken for an L0 layer. See the comment near `hack_end_key` (present in this diff but you need to fully expand the hidden files). It's nothing new, as `hack_image_layer_range` does the same on main.

If it were in a constant, it would be explained in an unified place.

File: pageserver/src/tenant/timeline.rs (hunk 4 above)
Comment (skyzh) [reply pageserver/src/tenant/timeline.rs:5455]: LSN ranges will overlap if there are single-key layers produced by legacy compaction. Maybe overlap is not the most accurate word here. See the figure below:

image

This is a valid layer map only when the first + third columns are single key layer files.

File: pageserver/src/tenant.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (skyzh) [reply pageserver/src/tenant.rs]: https://github.com/neondatabase/neon/pull/8608/commits/d6831d4c0feef7fc02d9b8af642aeda1f591eb58

**Comment by skyzh** [comment]
ready for review again :)

**Review by problame — COMMENTED** [review]
> The delta writer will keep the updates of the same key in a single file. This might create a super large layer, but we can optimize it later.

This is not a question about optimization but functional correctness.

We currently have hard limits on final delta layer file size due to the S3 upload file size limit: https://github.com/neondatabase/neon/blob/cfe493fa27567d2ce1688b8b570c4161a0644c78/pageserver/src/tenant/storage_layer/delta_layer.rs#L557-L565

If we repeatedly produce an over-sized layer, compaction will get stuck in retries / trip the circuit breaker.

That's ok risk for trial runs in staging, but a non-starter for trial-runs in prod. Please create a follow-up task and link it in the issue description + epic.

File: pageserver/src/tenant/storage_layer/split_writer.rs [hunk 8]
```diff
@@ -76,49 +108,108 @@ impl SplitImageLayerWriter {
...
+                drop(prev_image_writer);
+                self.generated_layers
+                    .push(SplitWriterResult::Discarded(layer_key));
+            } else {
+                self.generated_layers.push(SplitWriterResult::Produced(
+                    prev_image_writer
+                        .finish_with_end_key(tline, key, ctx)
+                        .await?,
+                ));
+            }
```
Comment (problame) [inline pageserver/src/tenant/storage_layer/split_writer.rs:128]: We drop the layer map lock inbetween discard check and finish_with_end_key.

It doesn't matter at this point because we have locked out both the GC and the Compaction background task (do we?).

File: pageserver/src/tenant/storage_layer/split_writer.rs [hunk 9]
```diff
@@ -76,49 +108,108 @@ impl SplitImageLayerWriter {
...
+        } else {
+            generated_layers.push(SplitWriterResult::Produced(
+                inner.finish_with_end_key(tline, end_key, ctx).await?,
+            ));
+        }
         Ok(generated_layers)
     }
 
-    /// When split writer fails, the caller should call this function and handle partially generated layers.
     #[allow(dead_code)]
```
Comment (problame) [inline pageserver/src/tenant/storage_layer/split_writer.rs:178]: This is not dead_code anymore

**Review by problame — COMMENTED** [review]
Why are we not deferring the "discard" hack until here: https://github.com/neondatabase/neon/blob/b0ca45cfdc130d8689a6c843c92e4297c1db2c7f/pageserver/src/tenant/timeline/compaction.rs#L2145-L2149 

The idea being:

1. don't call `.finish()` from the split writer
2. instead, queue up all writers into which we don't want to put more data
3. after that's all done
4. lock the layer map in write mode
5. for each queued split writer
  1. do the finish with discard check, against the locked layer map

Advantages:

This eliminates the need for the ugly `discard` async callback on the `pipe_to` path.

We eliminate the race condition (intermediately dropped that I pointed out earlier.

We get auto-cleanup both at runtime and on crash because unfinished writers have the Drop impl that does cleanup, and they use a temp suffix for the file name which we clean up on startup.

There are downsides, of course:
1. buffered writer's in-memory buffer stays around
2. finish() is not actually free, it still does IO
3. no way to evict the writers

(1) and (2) can be solved by refactoring the delta/image writers so the flow is `put put put finish_all_io convert_to_struct_layer` instead of the `put put put finish(finish_all_io, convert_to_struct_layer)` that we have today.

Wrt (3): in this PR, we don't evict the `struct Layers` that we collect in the split writer either. So, it's not a loss. And in the future, when we've removed the need for the hack and thus support having multiple `struct Layer` for the same persistent key but different files, we can do the convert_to_struct_layer early.

File: pageserver/src/tenant/storage_layer/split_writer.rs [hunk 10]
```diff
@@ -147,67 +238,140 @@ impl SplitDeltaLayerWriter {
...
+    ) -> anyhow::Result<()>
+    where
+        D: FnOnce(&PersistentLayerKey) -> F,
+        F: Future<Output = bool>,
+    {
         // The current estimation is key size plus LSN size plus value size estimation. This is not an accurate
         // number, and therefore the final layer size could be a little bit larger or smaller than the target.
+        //
+        // Also, keep all updates of a single key in a single file. TODO: split them using the legacy compaction
+        // strategy.
```
Comment (problame) [inline pageserver/src/tenant/storage_layer/split_writer.rs:264]: Please create a follow-up task in GH and link it here.

File: pageserver/src/tenant/timeline/compaction.rs [hunk 11]
```diff
@@ -1814,11 +1879,26 @@ impl Timeline {
...
             for desc in layers.iter_historic_layers() {
-                if desc.get_lsn_range().start <= gc_cutoff {
+                if desc.get_lsn_range().end <= max_layer_lsn {
                     selected_layers.push(guard.get_from_desc(&desc));
                 }
             }
+            if selected_layers.is_empty() {
+                info!("no layers to compact with gc");
+                return Ok(());
+            }
```
Comment (problame) [inline pageserver/src/tenant/timeline/compaction.rs:1901]: This looks like one of the optimizations to layer selection from the design doc.

They look correct but why is this optimization included in this PR? Seems like an orthogonal change (?)

Comment (problame) [reply pageserver/src/tenant/timeline/compaction.rs:1901]: Ohh, no, it's not a selection.

It's a change in behavior; it adds more layers to the selection than before.

Comment (problame) [reply pageserver/src/tenant/timeline/compaction.rs:1901]: image

Before this PR, we would not include C.

With this PR, we do include C.

Why the change?

File: pageserver/src/tenant/timeline/compaction.rs [hunk 12]
```diff
@@ -1848,25 +1928,49 @@ impl Timeline {
...
         );
         // Step 1: (In the future) construct a k-merge iterator over all layers. For now, simply collect all keys + LSNs.
-        // Also, collect the layer information to decide when to split the new delta layers.
+        // Also, verify if the layer map can be split by drawing a horizontal line at every LSN start/end split point.
+        let mut lsn_split_point = BTreeSet::new();
         let mut downloaded_layers = Vec::new();
-        let mut delta_split_points = BTreeSet::new();
         for layer in &layer_selection {
             let resident_layer = layer.download_and_keep_resident().await?;
             downloaded_layers.push(resident_layer);
```
Comment (problame) [inline pageserver/src/tenant/timeline/compaction.rs:1936]: Youy can push down the downloaded_layers business until after this `for` loop. 

This has the advantage of us not downloading a lot of layers, only to discover later that we can't run the bottom-most compaction because the layer map doesn't have the LSN split property.

File: pageserver/src/tenant/timeline/compaction.rs [hunk 13]
```diff
@@ -1848,25 +1928,49 @@ impl Timeline {
             lowest_retain_lsn
         );
         // Step 1: (In the future) construct a k-merge iterator over all layers. For now, simply collect all keys + LSNs.
-        // Also, collect the layer information to decide when to split the new delta layers.
+        // Also, verify if the layer map can be split by drawing a horizontal line at every LSN start/end split point.
+        let mut lsn_split_point = BTreeSet::new();
```
Comment (problame) [inline pageserver/src/tenant/timeline/compaction.rs:1932]:
```suggestion
        let mut lsn_split_point = BTreeSet::new(); // TODO: use a better data structure (range tree / range set?)
```

**Review by problame — APPROVED** [review]
Nice cleanup! 

I feel strongly about some of the comments, others are more observational. Hitting approve to unblock you, but please do follow up where requested.

File: pageserver/src/tenant/timeline.rs (hunk 4 above)
Comment (arpad-m) [reply pageserver/src/tenant/timeline.rs:5455]: but we had this case before as well, no? Why do we hit the panic now?

**Comment by skyzh** [comment]
> We get auto-cleanup both at runtime and on crash because unfinished writers have the Drop impl that does cleanup, and they use a temp suffix for the file name which we clean up on startup.

Yes I agree -- I created a follow-up item in the epic issue that we make the split layer writer atomic. This would be some amount of work but will make the split layer writer easier to use.

File: pageserver/src/tenant.rs [hunk 14]
```diff
@@ -7013,10 +7013,10 @@ mod tests {
                     lsn_range: Lsn(0x30)..Lsn(0x31),
                     is_delta: false
                 },
-                // The delta layer that is cut in the middle
+                // The delta layer covers the full range (with the layer key hack to avoid being recognized as L0)
                 PersistentLayerKey {
-                    key_range: get_key(3)..get_key(4),
-                    lsn_range: Lsn(0x30)..Lsn(0x41),
+                    key_range: Key::MIN..Key::NON_L0_MAX,
+                    lsn_range: Lsn(0x30)..Lsn(0x48),
```
[the remark this answers is not in the document]
Comment (skyzh) [reply pageserver/src/tenant.rs:7019]: Yes, I would create a follow up pull request to fix both image layer + delta layer ranges in gc-compaction. Image layer can cover full range now b/c the L0 criteria gets fixed, and delta layer should only cover the keys in the range.

File: pageserver/src/tenant/storage_layer/split_writer.rs (hunk 8 above)
Comment (skyzh) [reply pageserver/src/tenant/storage_layer/split_writer.rs:128]: We have GC+compaction lock so it doesn't matter.

File: pageserver/src/tenant/storage_layer/split_writer.rs (hunk 10 above)
Comment (skyzh) [reply pageserver/src/tenant/storage_layer/split_writer.rs:264]: https://github.com/neondatabase/neon/issues/8837

File: pageserver/src/tenant.rs (hunk 14 above)
[the remark this answers is not in the document]
Comment (skyzh) [reply pageserver/src/tenant.rs:7019]: https://github.com/neondatabase/neon/issues/8838

File: pageserver/src/tenant/timeline/compaction.rs (hunk 11 above)
Comment (skyzh) [reply pageserver/src/tenant/timeline/compaction.rs:1901]: Firstly, this is not a valid layer map if C contains multiple keys.

The change is made to consider the result of the legacy compaction, where it will split a delta layer containing a single key into multiple layers on the LSN horizon. To avoid creating delta layers overlapping with original delta layers in this case, we should take all layers below some LSN into the compaction process, instead of only the ones intersecting with the GC horizon.

File: pageserver/src/tenant/storage_layer/split_writer.rs (hunk 9 above)
Comment (skyzh) [reply pageserver/src/tenant/storage_layer/split_writer.rs]: This is still dead_code b/c we use finish_with_discard. The dead_code cfg attribute can be removed once I implement the atomic finish thing for the split layer writer.

File: pageserver/src/tenant/timeline/compaction.rs (hunk 11 above)
Comment (problame) [reply pageserver/src/tenant/timeline/compaction.rs:1901]: >  To avoid creating delta layers overlapping with original delta layers in this case, we should take all layers below some LSN into the compaction process, instead of only the ones intersecting with the GC horizon.

Ahh, makes sense. Maybe worth a comment? Feel free to resolve

File: libs/pageserver_api/src/key.rs [hunk 15]
```diff
@@ -236,6 +236,14 @@ impl Key {
         field5: u8::MAX,
         field6: u32::MAX,
     };
+    pub const NON_L0_MAX: Key = Key {
```
Comment (arpad-m) [inline libs/pageserver_api/src/key.rs:239]:
```suggestion
    /// A key slightly smaller than [`Key::MAX`] for use in layer key ranges to avoid them to be confused with L0 layers
    pub const NON_L0_MAX: Key = Key {
```

**Review by arpad-m — APPROVED** [review]

**Comment by arpad-m** [comment]
Also, could you make it use `NON_L0_MAX` in the one place in `tenant.rs` where it currently does `key.field6 -= 1;`? Then all locations that rely on the hack use that constant.

File: pageserver/src/tenant/timeline/compaction.rs (hunk 11 above)
Comment (skyzh) [reply pageserver/src/tenant/timeline/compaction.rs:1901]: added a comment in the codebase

**Comment by skyzh** [comment]
> Then all locations that rely on the hack use that constant.

I did that before that but just now I realized I didn't replace all places, and I pushed a fix :)
